### PR TITLE
add architecture doc

### PR DESCRIPTION
**Pull request by minrk**
A high-level reference for the architecture of nbconvert

One goal, as requested by @amueller, is to point people in the right direction when they want to know how to change something about nbconvert. This doc can accumulate links to examples/tutorials for custom preprocessors, templates, exporters, etc.

cc @willingc, @amueller

Also cc @jdfreder who has made a figure illustrating this before (though it may have vanished),
and @cameronoelsen who might want to take a stab at a new diagram.

### CODE REVIEW

**Comment by amueller** [comment]
you guys are awesome :)

**Comment by Carreau** [comment]
Objection to merge that and iterate ?

**Comment by jdfreder** [comment]
> Objection to merge that and iterate ?

Probably the best way to continue!  We can discuss here even though the PR is merged

File: docs/source/architecture.rst [hunk 1]
```diff
@@ -0,0 +1,123 @@
+.. _architecture:
+
+=========================
+Architecture of nbconvert
+=========================
+
+This is a high-level outline of the structure and objects in nbconvert,
+and how they are used in the pipeline of converting a notebook to any given fromat.
```
Comment (jdfreder) [inline docs/source/architecture.rst]: I read fromat as "Fro Matt" ;)

**Pull request merged**